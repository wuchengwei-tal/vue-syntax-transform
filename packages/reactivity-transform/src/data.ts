export enum ConstantTypes {
  NOT_CONSTANT = 0,
  CAN_SKIP_PATCH = 1,
  CAN_HOIST = 2,
  CAN_STRINGIFY = 3
}

export enum BindingTypes {
  /**
   * returned from data()
   */
  DATA = 'data',
  /**
   * declared as a prop
   */
  PROPS = 'props',
  /**
   * a local alias of a `<script setup>` destructured prop.
   * the original is stored in __propsAliases of the bindingMetadata object.
   */
  PROPS_ALIASED = 'props-aliased',
  /**
   * a let binding (may or may not be a ref)
   */
  SETUP_LET = 'setup-let',
  /**
   * a const binding that can never be a ref.
   * these bindings don't need `unref()` calls when processed in inlined
   * template expressions.
   */
  SETUP_CONST = 'setup-const',
  /**
   * a const binding that does not need `unref()`, but may be mutated.
   */
  SETUP_REACTIVE_CONST = 'setup-reactive-const',
  /**
   * a const binding that may be a ref.
   */
  SETUP_MAYBE_REF = 'setup-maybe-ref',
  /**
   * bindings that are guaranteed to be refs
   */
  SETUP_REF = 'setup-ref',
  /**
   * declared by other options, e.g. computed, inject
   */
  OPTIONS = 'options'
}

export enum NodeTypes {
  ROOT = 0,
  ELEMENT = 1,
  TEXT = 2,
  COMMENT = 3,
  SIMPLE_EXPRESSION = 4,
  INTERPOLATION = 5,
  ATTRIBUTE = 6,
  DIRECTIVE = 7,
  COMPOUND_EXPRESSION = 8,
  IF = 9,
  IF_BRANCH = 10,
  FOR = 11,
  TEXT_CALL = 12,
  VNODE_CALL = 13,
  JS_CALL_EXPRESSION = 14,
  JS_OBJECT_EXPRESSION = 15,
  JS_PROPERTY = 16,
  JS_ARRAY_EXPRESSION = 17,
  JS_FUNCTION_EXPRESSION = 18,
  JS_CONDITIONAL_EXPRESSION = 19,
  JS_CACHE_EXPRESSION = 20,
  JS_BLOCK_STATEMENT = 21,
  JS_TEMPLATE_LITERAL = 22,
  JS_IF_STATEMENT = 23,
  JS_ASSIGNMENT_EXPRESSION = 24,
  JS_SEQUENCE_EXPRESSION = 25,
  JS_RETURN_STATEMENT = 26
}

// NodeTypes
export const ROOT = 0
export const ELEMENT = 1
export const TEXT = 2
export const COMMENT = 3
export const SIMPLE_EXPRESSION = 4
export const INTERPOLATION = 5
export const ATTRIBUTE = 6
export const DIRECTIVE = 7
export const COMPOUND_EXPRESSION = 8
export const IF = 9
export const IF_BRANCH = 10
export const FOR = 11
export const TEXT_CALL = 12
export const VNODE_CALL = 13
export const JS_CALL_EXPRESSION = 14
export const JS_OBJECT_EXPRESSION = 15
export const JS_PROPERTY = 16
export const JS_ARRAY_EXPRESSION = 17
export const JS_FUNCTION_EXPRESSION = 18
export const JS_CONDITIONAL_EXPRESSION = 19
export const JS_CACHE_EXPRESSION = 20
export const JS_BLOCK_STATEMENT = 21
export const JS_TEMPLATE_LITERAL = 22
export const JS_IF_STATEMENT = 23
export const JS_ASSIGNMENT_EXPRESSION = 24
export const JS_SEQUENCE_EXPRESSION = 25
export const JS_RETURN_STATEMENT = 26

// ConstantTypes
export const NOT_CONSTANT = 0
export const CAN_SKIP_PATCH = 1
export const CAN_HOIST = 2
export const CAN_STRINGIFY = 3

// BindingTypes
export const DATA = 'data'
export const PROPS = 'props'
export const PROPS_ALIASED = 'props-aliased'
export const SETUP_LET = 'setup-let'
export const SETUP_CONST = 'setup-const'
export const SETUP_REACTIVE_CONST = 'setup-reactive-const'
export const SETUP_MAYBE_REF = 'setup-maybe-ref'
export const SETUP_REF = 'setup-ref'
export const OPTIONS = 'options'
