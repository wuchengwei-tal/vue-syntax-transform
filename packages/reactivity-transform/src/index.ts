export { compileScript } from './compileScript'

import { compileScript } from './compileScript'

export { SFCScriptCompileOptions } from './compileScript'

import { parse } from '@vue/compiler-sfc'

export function reactivityTransform(src: string, id = 'xxxxxx') {
  const { descriptor } = parse(src)
  return compileScript(descriptor, {
    id,
    reactivityTransform: true
  })
}
