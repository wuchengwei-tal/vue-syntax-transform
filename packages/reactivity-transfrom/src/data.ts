export enum ConstantTypes {
  NOT_CONSTANT = 0,
  CAN_SKIP_PATCH = 1,
  CAN_HOIST = 2,
  CAN_STRINGIFY = 3,
}

export enum BindingTypes {
  /**
   * returned from data()
   */
  DATA = "data",
  /**
   * declared as a prop
   */
  PROPS = "props",
  /**
   * a local alias of a `<script setup>` destructured prop.
   * the original is stored in __propsAliases of the bindingMetadata object.
   */
  PROPS_ALIASED = "props-aliased",
  /**
   * a let binding (may or may not be a ref)
   */
  SETUP_LET = "setup-let",
  /**
   * a const binding that can never be a ref.
   * these bindings don't need `unref()` calls when processed in inlined
   * template expressions.
   */
  SETUP_CONST = "setup-const",
  /**
   * a const binding that does not need `unref()`, but may be mutated.
   */
  SETUP_REACTIVE_CONST = "setup-reactive-const",
  /**
   * a const binding that may be a ref.
   */
  SETUP_MAYBE_REF = "setup-maybe-ref",
  /**
   * bindings that are guaranteed to be refs
   */
  SETUP_REF = "setup-ref",
  /**
   * declared by other options, e.g. computed, inject
   */
  OPTIONS = "options",
}

export enum NodeTypes {
  ROOT = 0,
  ELEMENT = 1,
  TEXT = 2,
  COMMENT = 3,
  SIMPLE_EXPRESSION = 4,
  INTERPOLATION = 5,
  ATTRIBUTE = 6,
  DIRECTIVE = 7,
  COMPOUND_EXPRESSION = 8,
  IF = 9,
  IF_BRANCH = 10,
  FOR = 11,
  TEXT_CALL = 12,
  VNODE_CALL = 13,
  JS_CALL_EXPRESSION = 14,
  JS_OBJECT_EXPRESSION = 15,
  JS_PROPERTY = 16,
  JS_ARRAY_EXPRESSION = 17,
  JS_FUNCTION_EXPRESSION = 18,
  JS_CONDITIONAL_EXPRESSION = 19,
  JS_CACHE_EXPRESSION = 20,
  JS_BLOCK_STATEMENT = 21,
  JS_TEMPLATE_LITERAL = 22,
  JS_IF_STATEMENT = 23,
  JS_ASSIGNMENT_EXPRESSION = 24,
  JS_SEQUENCE_EXPRESSION = 25,
  JS_RETURN_STATEMENT = 26,
}
