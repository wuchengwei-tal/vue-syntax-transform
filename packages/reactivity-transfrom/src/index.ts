// API
export { parse } from "./parse";
export { compileTemplate } from "./compileTemplate";
export { compileScript } from "./compileScript";
export { rewriteDefault } from "./rewriteDefault";
export {
    shouldTransform as shouldTransformRef,
    transform as transformRef,
    transformAST as transformRefAST,
} from "@vue/reactivity-transform";

// Utilities
export { parse as babelParse } from "@babel/parser";
import MagicString from "magic-string";
export { MagicString };
// technically internal but we want it in @vue/repl, cast it as any to avoid
// relying on estree types
import { walk as _walk } from "estree-walker";
export const walk = _walk as any;

// Types
export {
    SFCParseOptions,
    SFCParseResult,
    SFCDescriptor,
    SFCBlock,
    SFCTemplateBlock,
    SFCScriptBlock,
    SFCStyleBlock,
} from "./parse";
export {
    TemplateCompiler,
    SFCTemplateCompileOptions,
    SFCTemplateCompileResults,
} from "./compileTemplate";

export { SFCScriptCompileOptions } from "./compileScript";
export {
    AssetURLOptions,
    AssetURLTagConfig,
} from "./templateTransformAssetUrl";
