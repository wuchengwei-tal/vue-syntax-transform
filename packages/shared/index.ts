export * from './warn'
export * from './test-utils'
export * from './cache'
export * from './const'
export * from './transform'

export const DEFAULT_FILENAME = 'anonymous.vue'

// Special compiler macros
export const DEFINE_PROPS = 'defineProps'
export const DEFINE_EMITS = 'defineEmits'
export const DEFINE_EXPOSE = 'defineExpose'
export const WITH_DEFAULTS = 'withDefaults'

// constants
export const DEFAULT_VAR = `__default__`
