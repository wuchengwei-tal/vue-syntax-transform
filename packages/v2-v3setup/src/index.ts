import {
  type SFCDescriptor,
  parseComponent,
  compile
} from 'vue-template-compiler'

import { compileScript } from './compileScript'

import { templateTransform } from './template-transform'
import { cssTransform } from './css-transform'
import MagicString from 'magic-string'

export * from './data'

type TransformedSFC = {
  template?: ReturnType<typeof templateTransform>
  script?: ReturnType<typeof compileScript>
  styles?: ReturnType<typeof cssTransform>[]
  content: string
}

export function sfcTransform(sfc: string, id = '') {
  const descriptor = parseComponent(sfc) as SFCDescriptor & {
    source: string
    cssVars: string[]
    errors: any[]
  }

  const { script, template, styles, source } = descriptor

  const s = new MagicString(source)

  const ret: TransformedSFC = { content: '', styles: [] }

  let inheritAttrs = true
  if (script) {
    // console.log(descriptor)
    const match = script.content.match(/inheritAttrs\s*:\s*(true|false)/)
    if (match) inheritAttrs = match[1] === 'true'
  }

  if (template) {
    const { content, start, end } = template

    const { ast } = compile(content, {
      outputSourceRange: true,
      whitespace: 'preserve'
    })
    ret.template = templateTransform({ ast, ...template, inheritAttrs })
    s.overwrite(start!, end!, ret.template.content)
  }

  if (script) {
    const { start, end, content } = script
    if (content) {
      ret.script = compileScript(content, { id })
      s.overwrite(start!, end!, '\n' + ret.script.content + '\n')
      let i = start!
      while (source[i--] === '>');
      while (source[--i] !== '<');
      if (source.slice(i, i + 7) === '<script') {
        s.prependRight(i + 7, ' setup')
      }
    }
  }

  if (styles.length) {
    styles.forEach(style => {
      const { start, end, content } = style
      if (content) {
        const res = cssTransform(content)
        ret.styles?.push(res)
        s.overwrite(start!, end!, res)
      }
    })
  }

  return { ...ret, content: s.toString() }
}

function transformActions(code: string) {
  const fnRe = /(export )?const \w+ = (async )?\(.*[\)|\n]/g

  let match
  let last = 0
  let output = ''
  while ((match = fnRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [line, _, __] = match

    last = match.index + line.length

    let flag = line[line.length - 2]
    // console.log(line);
    if (line) {
      // oneline params
      if (flag === '{') {
        const match = /\(((\{.*\})|(\w+)),? ?(.*)?\)/.exec(line)
        if (match) {
          let [arg1, _, __, ___, arg2e] = match
          line = line.replace(arg1.slice(1, -1), arg2e || '')
        }
      }
      output += pre + line
      // multi line params
      if (flag === '(') {
        const params = parseInside(code, last, '(')
        let i = 0
        while (params[i++] !== '}');
        if (params[i] === ',') i++
        output += params.slice(i)

        last += params.length
        while (code[last++] !== '{') output += code[last]
        flag = '{'
      }
    }

    if (flag === '{') {
      let body = parseInside(code, last, '{')
      last += body.length
      body = replaceActionBody(body)
      body = replaceState(body)
      output += body
    }
  }
  output += code.slice(last)

  // console.log(output);
  return output
}

const Pair = { '{': '}', '(': ')', '[': ']' }
function parseInside(code: string, s: number, p: keyof typeof Pair) {
  let e = s
  let cnt = 0
  while (e < code.length - 1) {
    const ch = code[e++]
    if (ch === p) cnt++
    if (ch === Pair[p]) {
      if (cnt) cnt--
      else break
    }
  }
  const body = code.slice(s, e)
  return body
}

function replaceActionBody(code: string) {
  const dispatchRe = /(((store\.)?dispatch)|(commit))\('?((\w+\.?\w+))'?,? ?/g

  let match
  let last = 0
  let output = ''
  while ((match = dispatchRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [line, _, store, dispatch, commit, fn] = match

    last = match.index + line.length

    if (line && fn) {
      if (fn.includes('.')) fn = fn.split('.')[1]
      line = fn + '('
      // console.log(line)
    }

    output += pre + line
  }
  output += code.slice(last)
  // console.log(output);
  return output
}

function transformMutations(code: string) {
  const { body, index } = getExportDefault(code)
  if (!body) return code

  let last = index
  let output = code.slice(0, index)

  let match
  const mutationRe = /((\w+)|\[types.(\w+)\])(\(state,?.*\)) {/g
  while ((match = mutationRe.exec(body))) {
    let pre = ''
    if (last !== index) pre = body.slice(last, match.index)

    let [line, _, name, type, params] = match

    last = match.index + line.length
    params = params.replace(/state,? ?/, '')
    line = 'export function ' + (name || type) + params + '{'

    let mutationBody = parseInside(body, last, '{')
    last += mutationBody.length + 1 // del comma

    mutationBody = replaceState(mutationBody)

    output += pre + line + mutationBody
  }

  return output
}

function getExportDefault(code: string) {
  const defaultRe = /export default {/
  const match = defaultRe.exec(code)
  if (match) {
    return {
      body: parseInside(code, match.index, '{'),
      index: match.index
    }
  }
  return {}
}

function replaceState(code: string) {
  const stateRe = /(state\.\w+)( |\(|\.|\[|,)/g
  let last = 0
  let output = ''
  let match
  while ((match = stateRe.exec(code))) {
    const pre = code.slice(last, match.index)

    let [res, state] = match
    last = match.index + res.length

    if (res.endsWith('.')) res = res.replace(res, `${res}value.`)
    else res = res.replace(state, `${state}.value`)
    res = res.replace('state.', '')

    output += pre + res
  }
  output += code.slice(last)

  output = output.replace(/rootState\??\.\w+\??\./g, '')

  return output
}
