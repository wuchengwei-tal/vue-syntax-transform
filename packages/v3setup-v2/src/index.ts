import { parse } from '@vue/compiler-sfc'

import { compileScript } from './compileScript'
import MagicString from 'magic-string'

type TransformedSFC = {
  // template?: ReturnType<typeof templateTransform>
  script?: ReturnType<typeof compileScript>
  styles?: ReturnType<any>[]
  content: string
}

export function v3SetupToV2(sfc: string, id = '') {
  const { descriptor } = parse(sfc)
  const { script, scriptSetup, template, styles, source } = descriptor

  const s = new MagicString(source)

  const ret: TransformedSFC = { content: '', styles: [] }

  if (script || scriptSetup) {
    ret.script = compileScript(descriptor, { id: '', hoistStatic: true })

    let start = -1
    let end = -1
    if (scriptSetup) {
      start = scriptSetup.loc.start.offset
      end = scriptSetup.loc.end.offset
    }
    if (script) {
      start = start!
        ? Math.min(start, script.loc.start.offset)
        : script.loc.start.offset
      end = end! ? Math.max(end, script.loc.end.offset) : script.loc.end.offset
    }
    if (start !== -1) {
      s.overwrite(start, end, '\n' + ret.script.content + '\n')

      let i = start!
      let len = 0
      while (source[i--] === '>');
      while (source[--i] !== '<') len++
      s.overwrite(i, i + len + 1, '<script')
    }
  }

  return { ...ret, content: s.toString() }
}

export function transform(code: string) {
  const { descriptor } = parse(code)
  return compileScript(descriptor, {
    id: '',
    hoistStatic: true,
    reactivityTransform: true
  })
}
