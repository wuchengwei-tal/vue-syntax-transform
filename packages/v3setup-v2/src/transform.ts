import {
  BindingMap,
  BindingTypes,
  isEmptyStmt,
  isMember
} from '@vue-transform/shared'

import {
  Expression,
  ClassDeclaration,
  FunctionDeclaration,
  Identifier,
  CallExpression,
  ObjectProperty,
  ObjectMethod,
  ObjectExpression,
  BlockStatement,
  Function,
  Statement,
  Node,
  MemberExpression,
  //
  objectProperty,
  identifier,
  stringLiteral,
  objectExpression,
  objectMethod,
  blockStatement,
  returnStatement,
  expressionStatement,
  memberExpression,
  isFunction
} from '@babel/types'

import { walk } from 'estree-walker'

const generate = require('@babel/generator').default

export type BindingValue =
  | Expression
  | ClassDeclaration
  | FunctionDeclaration
  | Identifier
  | CallExpression['arguments']

export type TransformBindingsMap = BindingMap<BindingValue> & {
  watch: BindingMap<BindingValue>
  $hooks: BindingMap<BindingValue>
  $refs: BindingMap<null>
}

type Options = {
  data: ObjectProperty[]
  computed: (ObjectMethod | ObjectProperty)[]
  methods: ObjectMethod[]
  watch: ObjectProperty[]
  hooks: ObjectMethod[]
}

export function transformBindings(bindings: TransformBindingsMap) {
  const options: Options = {
    data: [],
    computed: [],
    methods: [],
    watch: [],
    hooks: []
  }

  function transBody(value: BlockStatement | Expression): BlockStatement {
    // @ts-expect-error
    walk(value, {
      enter(child: Node, parent: Node) {
        if (isMember(child)) {
          if (
            child.object.type === 'Identifier' &&
            child.object.name in bindings
          ) {
            let _this: Identifier | MemberExpression = identifier('this')
            if (child.object.name in bindings.$refs) {
              _this = memberExpression(_this, identifier('$refs'))
            }
            if (
              child.property.type === 'Identifier' &&
              child.property.name === 'value'
            ) {
              if (isMember(parent)) {
                child.property.name = child.object.name
                child.object = _this
              } else {
                child.property = identifier(child.object.name)
                child.object = _this
              }
            } else {
              child.object = memberExpression(
                _this,
                identifier(child.object.name)
              )
            }
          }
        }

        if (child.type === 'Identifier') {
          if (['router', 'route', 'emit'].includes(child.name)) {
            child.name = 'this.$' + child.name
          }
          if (['props', '__props'].includes(child.name)) {
            child.name = 'this'
          }
        }
      }
    })

    if (value.type === 'BlockStatement') {
      value.body = value.body.filter(v => !isEmptyStmt(v))
      return value
    } else {
      const stmt = expressionStatement(value)
      return blockStatement(isEmptyStmt(stmt) ? [] : [stmt])
    }
  }

  function normalizeBody(
    node: Function,
    statement: (expr: Expression) => Statement = expressionStatement
  ) {
    let body: BlockStatement | undefined
    if (node.type === 'ArrowFunctionExpression') {
      if (node.body.type === 'BlockStatement') body = node.body
      else body = blockStatement([statement(node.body)])
    }
    if (
      node.type === 'FunctionExpression' ||
      node.type === 'FunctionDeclaration'
    ) {
      body = node.body
    }
    if (body) body = transBody(body)

    return body
  }

  function transWatcher(key: string, value: BindingValue, options: Options) {
    if (!Array.isArray(value)) return

    const [cb, opts] = value
    if (isFunction(cb)) {
      const body = normalizeBody(cb)
      if (body) {
        const handler = objectMethod(
          'method',
          identifier('handler'),
          transParams(cb.params),
          body
        )

        options.watch.push(
          objectProperty(
            key.includes('.') ? stringLiteral(key) : identifier(key),
            objectExpression([
              handler,
              ...(opts as ObjectExpression).properties
            ])
          )
        )
      }
    }
  }

  function transHook(key: string, value: BindingValue, options: Options) {
    if (!Array.isArray(value)) return
    if (!isFunction(value[0])) return

    const body = normalizeBody(value[0])

    if (body) {
      const params = value[0].params

      options.hooks.push(
        objectMethod(
          'method',
          identifier(key),
          transParams(params),
          body,
          false,
          value[0].generator,
          value[0].async
        )
      )
    }
  }

  function transMethod(key: string, value: BindingValue, options: Options) {
    if (!isFunction(value)) return

    const body = normalizeBody(value)

    if (body) {
      options.methods.push(
        objectMethod('method', identifier(key), transParams(value.params), body)
      )
    }
  }

  function transGetters(key: string, value: BindingValue, options: Options) {
    if (!Array.isArray(value)) return
    if (!isFunction(value[0])) return
    const body = normalizeBody(value[0], returnStatement)

    if (body) {
      options.computed.push(objectMethod('method', identifier(key), [], body))
    }
  }

  for (let [key, { type, value }] of Object.entries(bindings)) {
    if (type === BindingTypes.DATA)
      transState(key, value as BindingValue, options)
    if (type === BindingTypes.COMPUTED)
      transGetters(key, value as BindingValue, options)
    if (type === BindingTypes.METHOD)
      transMethod(key, value as BindingValue, options)

    if (key === 'watch') {
      for (const [key, { value }] of Object.entries(bindings.watch))
        transWatcher(key, value, options)
    }
    if (key === '$hooks') {
      for (const [key, { value }] of Object.entries(bindings.$hooks))
        transHook(key, value, options)
    }
  }

  const output = []

  if (options.data.length) {
    output.push(
      objectMethod(
        'method',
        identifier('data'),
        [],
        blockStatement([returnStatement(objectExpression(options.data))])
      )
    )
  }

  if (options.computed.length) {
    output.push(
      objectProperty(identifier('computed'), objectExpression(options.computed))
    )
  }

  if (options.methods.length) {
    output.push(
      objectProperty(identifier('methods'), objectExpression(options.methods))
    )
  }

  if (options.watch.length) {
    output.push(
      objectProperty(identifier('watch'), objectExpression(options.watch))
    )
  }

  output.push(...options.hooks)

  const code: string = output.length
    ? generate(objectExpression(output)).code || ''
    : ''
  const i = code.indexOf('{')
  const j = code.lastIndexOf('}')
  return code.slice(i + 1, j)
}

function transState(key: string, value: BindingValue, options: Options) {
  if (Array.isArray(value)) {
    if (!value.length) value[0] = identifier('undefined')
    const { type } = value[0]
    if (
      type === 'SpreadElement' ||
      type === 'JSXNamespacedName' ||
      type === 'ArgumentPlaceholder'
    )
      return
    options.data.push(objectProperty(identifier(key), value[0]))
  } else {
    const { type } = value
    if (type === 'FunctionDeclaration' || type === 'ClassDeclaration') return
    options.data.push(objectProperty(identifier(key), value))
  }
}

function transParams(params: Function['params']) {
  for (const param of params) {
    if ('typeAnnotation' in param) {
      param.typeAnnotation = null
    }
    if ('optional' in param) {
      param.optional = false
    }
  }
  return params as ObjectMethod['params']
}
