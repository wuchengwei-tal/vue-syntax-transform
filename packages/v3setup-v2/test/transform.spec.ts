import { BindingTypes, assertCode } from '@vue-transform/shared'
import { transform } from '../src'

describe('script transform', () => {
  test('ref', () => {
    const { content, transBindings } = transform(`
<script lang="ts" setup>
const undef = ref();
const nickname = ref<string>('nickname');
const avatar = ref<{img:string}>({img: 'avatar.png'});
const visible = ref<boolean>(false);
const data = reactive<{a:number}>({a:1})
let a = reactive<{a:number}>({a:1})
let b = 1

avatar.value.img = 'new avatar.png'
</script>
`)
    assertCode(content)

    if (!transBindings) return
    expect(transBindings['nickname']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['avatar']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['visible']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['data']?.type).toEqual(BindingTypes.DATA)
  })

  test('macros', () => {
    const { content } = transform(`
<script lang="ts" setup>
defineEmits(['change']);
defineProps<{a?:number}>()
defineExpose({a:1})
defineOptions({
    name: 'Component',
    inheritAttrs: false,
})
</script>
`)
    assertCode(content)

    expect(content).not.toMatch('defineEmits')
    expect(content).not.toMatch('defineProps')
    expect(content).not.toMatch('defineExpose')
    expect(content).not.toMatch('defineOptions')
    expect(content).not.toMatch('expose')
  })

  test('imports', () => {
    const { content } = transform(`
<script lang="ts" setup>
import img from './img.webp';
import Component from './Component.vue';
import type { CheckboxChangeEvent } from 'ant-design-vue';
import { type CheckboxGroupProps, Checkbox } from 'ant-design-vue';
</script>
`)
    assertCode(content)
  })

  test('computed', () => {
    const { content, transBindings } = transform(`
<script lang="ts" setup>
const nickname = ref('nickname');
const name = computed(() => nickname.value);
const age = computed(() => 18);
</script>
`)
    assertCode(content)

    if (!transBindings) return
    expect(transBindings['nickname']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['name']?.type).toEqual(BindingTypes.COMPUTED)
    expect(transBindings['age']?.type).toEqual(BindingTypes.COMPUTED)
  })
  test('methods', () => {
    const { content, transBindings } = transform(`
<script lang="ts" setup>
const nickname = ref('nickname');
const name = computed(() => nickname.value);
const age = computed(() => 18);

function changeName(a?: any) {
  nickname.value = 'new nickname'
}

const changeAge = (b? :any) => {
  age.value = 20
}
</script>
`)
    assertCode(content)

    if (!transBindings) return
    expect(transBindings['nickname']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['name']?.type).toEqual(BindingTypes.COMPUTED)
    expect(transBindings['age']?.type).toEqual(BindingTypes.COMPUTED)
    expect(transBindings['changeName']?.type).toEqual(BindingTypes.METHOD)
    expect(transBindings['changeAge']?.type).toEqual(BindingTypes.METHOD)
  })

  test('watch', () => {
    const { content, transBindings } = transform(`
<script lang="ts" setup>
const nickname = ref('nickname');
const info = reactive({a:1})
const name = computed(() => nickname.value);
const age = computed(() => info.a);

watch(nickname, () => {
  console.log('a changed')
  info.a++
}, { immediate: true, deep: true })

watch(() => info.a, () => {
  console.log('a changed')
  info.a++
}, { immediate: true, deep: true })

watch(() => age, () => {
  console.log('a changed')
  info.a++
}, { immediate: true, deep: true })

watchEffect(()=> {
  info.a++
  console.log('a changed')
})
</script>
`)
    assertCode(content)

    if (!transBindings) return
    expect(transBindings['nickname']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['info']?.type).toEqual(BindingTypes.DATA)
    expect(transBindings['name']?.type).toEqual(BindingTypes.COMPUTED)
    expect(transBindings['age']?.type).toEqual(BindingTypes.COMPUTED)
    expect(transBindings['watch']['nickname']?.type).toEqual(BindingTypes.WATCH)
    expect(transBindings['watch']['age']?.type).toEqual(BindingTypes.WATCH)
    expect(transBindings['watch']['info.a']?.type).toEqual(BindingTypes.WATCH)
    // expect(transBindings['watchEffect']?.type).toEqual(BindingTypes.WATCH)
  })
  test('life cycles hooks', () => {
    const { content } = transform(`
<script lang="ts" setup>
const a = ref({c: 1})
onMounted(()=> {
  a.value.c = 2
})

onUnmounted(()=> {
  a.value.c = 3
})
</script>
`)
    assertCode(content)
  })

  test('cssVars', () => {
    const { content } = transform(`
<script lang="ts" setup>
const a = ref('1px')
const b = ref(2)
</script>

<style >
div{
  width: v-bind(a);
  height: v-bind(b);
}
</style>
`)
    assertCode(content)
  })

  test('emit, router', () => {
    const { content } = transform(`
<script lang="ts" setup>
const route = useRoute()
const router = useRouter()
const path = computed(() => route.path)
const emit = defineEmits(['change'])


onMounted(()=> {
  emit('change', 1)
  router.push({path: '/a'})
  router.go(1)
})
emit('change', 2)
router.replace({path: '/a'})
</script>
`)

    assertCode(content)
  })

  test('props', () => {
    const { content } = transform(`
<script lang="ts" setup>
const props = defineProps<{a?:number}>()

const a = computed(()=> props.a)
console.log(props.a)
`)
    assertCode(content)
  })

  test('types', () => {
    const { content } = transform(`
<script lang="ts" setup>
type A = {
  a: number
}
interface B {
  b: string
}

enum Enum {
  A = 'a',
}
</script>
`)

    assertCode(content)
  })

  test('template reference', () => {
    const { content } = transform(`
<script lang="ts" setup>
const divRef = ref<HTMLDivElement>()
const spanRef = ref<HTMLDivElement|null>(null)

onMounted(()=> {
  divRef.value?.focus()
  spanRef.value?.click()
})
</script>

<template>
  <div ref="divRef"></div>
  <span ref="spanRef"></span>
</template>
`)
    assertCode(content)
  })
})
