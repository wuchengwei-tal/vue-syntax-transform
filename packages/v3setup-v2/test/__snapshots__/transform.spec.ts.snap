// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`script transform > computed 1`] = `
"export default {
  
  data() {
    return {
      nickname: 'nickname'
    };
  },
  computed: {
    name() {
      return nickname.value;
    },
    age() {
      return 18;
    }
  }
,
  created() {




}}"
`;

exports[`script transform > cssVars 1`] = `
"export default {
  
  data() {
    return {
      a: '1px',
      b: 2
    };
  }
,
  created() {



}}"
`;

exports[`script transform > imports 1`] = `
"import img from './img.webp';
import Component from './Component.vue';
import { Checkbox } from 'ant-design-vue';

export default {
  created() {

}}"
`;

exports[`script transform > life cycles hooks 1`] = `
"export default {
  
  data() {
    return {
      a: 1
    };
  },
  mounted() {
    a.value = 2;
  },
  destroyed() {
    a.value = 3;
  }
,
  created() {





}}"
`;

exports[`script transform > macros 1`] = `
"export default {
  ...{
    name: 'Component',
    inheritAttrs: false,
},
  props: {
    a: { type: Number, required: false }
  },
  created() {





}}"
`;

exports[`script transform > methods 1`] = `
"export default {
  
  data() {
    return {
      nickname: 'nickname'
    };
  },
  computed: {
    name() {
      return nickname.value;
    },
    age() {
      return 18;
    }
  },
  methods: {
    changeName() {
      nickname.value = 'new nickname';
    },
    changeAge() {
      age.value = 20;
    }
  }
,
  created() {








}}"
`;

exports[`script transform > ref 1`] = `
"export default {
  
  data() {
    return {
      nickname: 'nickname',
      avatar: img,
      visible: false,
      data: {
        a: 1
      },
      a: {
        a: 1
      }
    };
  }
,
  created() {






let b = 1
}}"
`;

exports[`script transform > watch 1`] = `
"export default {
  
  data() {
    return {
      nickname: 'nickname',
      info: {
        a: 1
      }
    };
  },
  computed: {
    name() {
      return nickname.value;
    },
    age() {
      return info.a;
    }
  },
  watch: {
    nickname: {
      handler() {
        console.log('a changed');
        info.a++;
      },
      immediate: true,
      deep: true
    },
    \\"info.a\\": {
      handler() {
        console.log('a changed');
        info.a++;
      },
      immediate: true,
      deep: true
    },
    age: {
      handler() {
        console.log('a changed');
        info.a++;
      },
      immediate: true,
      deep: true
    }
  }
,
  created() {













}}"
`;
